### PR TITLE
feat: Implement storage extension FIP

### PR DESCRIPTION
**Pull request by sanjayprabhu**
## Why is this change needed?

Implements https://github.com/farcasterxyz/protocol/discussions/191

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] PR has a [changeset](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#35-adding-changesets)
- [x] PR has been tagged with a change label(s) (i.e. documentation, feature, bugfix, or chore)
- [ ] PR includes [documentation](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#32-writing-docs) if necessary.

---

## PR-Codex overview
The focus of this PR is to implement the Storage Extension FIP. 

### Detailed summary
- Added `StorageUnitDetails` message in `request_response.proto`
- Updated storage-related functions in various Rust files
- Modified default limits in different store types
- Introduced new date calculations in `factories.ts`

> The following files were skipped due to too many changes: `apps/hubble/src/storage/stores/storeEventHandler.test.ts`, `packages/core/src/limits.ts`, `apps/hubble/src/storage/stores/storageCache.test.ts`, `apps/hubble/src/storage/stores/storeEventHandler.ts`, `apps/hubble/src/rpc/test/server.test.ts`, `packages/core/src/limits.test.ts`, `apps/hubble/src/storage/engine/index.ts`, `packages/hub-web/src/generated/request_response.ts`, `packages/hub-nodejs/src/generated/request_response.ts`, `packages/core/src/protobufs/generated/request_response.ts`, `apps/hubble/src/storage/stores/storageCache.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/hubble/src/storage/engine/index.ts [hunk 1]
```diff
@@ -172,15 +173,14 @@ class Engine extends TypedEmitter<EngineEvents> {
...
-    // is not included in this calculation because it is not pruned.
+    // Total set size for all stores. We should probably reduce this to just the size of the cast store.
     this._totalPruneSize =
-      this._linkStore.pruneSizeLimit +
-      this._reactionStore.pruneSizeLimit +
-      this._castStore.pruneSizeLimit +
-      this._userDataStore.pruneSizeLimit +
-      this._verificationStore.pruneSizeLimit +
-      this._usernameProofStore.pruneSizeLimit;
+      getDefaultStoreLimit(StoreType.CASTS, StorageUnitType.UNIT_TYPE_LEGACY) +
```
Comment (aditiharini) [inline apps/hubble/src/storage/engine/index.ts:178]: Why are all these things in terms of UNIT_TYPE LEGACY? Will it always be that way?

File: apps/hubble/src/storage/stores/linkStore.ts [hunk 2]
```diff
@@ -38,7 +32,7 @@ import storeEventHandler from "./storeEventHandler.js";
  */
 class LinkStore extends RustStoreBase<LinkAddMessage, LinkRemoveMessage> {
   constructor(db: RocksDB, eventHandler: storeEventHandler, options: StorePruneOptions = {}) {
-    const pruneSizeLimit = options.pruneSizeLimit ?? getDefaultStoreLimit(StoreType.LINKS);
+    const pruneSizeLimit = options.pruneSizeLimit ?? 0;
```
Comment (aditiharini) [inline apps/hubble/src/storage/stores/linkStore.ts:35]: Why did this diff happen?

File: apps/hubble/src/storage/stores/storageCache.ts [hunk 3]
```diff
@@ -21,14 +23,44 @@ import { forEachOnChainEvent } from "../db/onChainEvent.js";
 import { addProgressBar } from "../../utils/progressBars.js";
 
 const MAX_PENDING_MESSAGE_COUNT_SCANS = 100;
+export const LEGACY_STORAGE_UNIT_CUTOFF_TIMESTAMP = 1724889600; // 2024-08-29 00:00:00 UTC
```
Comment (aditiharini) [inline apps/hubble/src/storage/stores/storageCache.ts:26]: Why is the time unit in seconds and not milliseconds? Is blockTimestamp in the same units?

File: apps/hubble/src/storage/stores/storageCache.ts [hunk 4]
```diff
@@ -65,15 +97,25 @@ export class StorageCache {
...
+          if (rentEventSlot.invalidateAt < time.value) {
+            return;
+          }
+
+          if (existingSlot) {
+            this._activeStorageSlots.set(event.fid, {
+              units: rentEventSlot.units + existingSlot.units,
+              legacy_units: rentEventSlot.legacy_units + existingSlot.legacy_units,
+              invalidateAt:
+                existingSlot.invalidateAt < rentEventSlot.invalidateAt
```
Comment (aditiharini) [inline apps/hubble/src/storage/stores/storageCache.ts:112]: Why are we picking the soonest invalidation time? The different slots will expire at different times depending on when they came in right?

File: apps/hubble/src/storage/stores/storageCache.ts [hunk 5]
```diff
@@ -21,14 +23,44 @@ import { forEachOnChainEvent } from "../db/onChainEvent.js";
...
+    units: isLegacy ? 0 : event.storageRentEventBody.units,
+    legacy_units: isLegacy ? event.storageRentEventBody.units : 0,
+    invalidateAt: toFarcasterTime(getStorageUnitExpiry(event) * 1000).unwrapOr(0),
+  };
+};
+
 const log = logger.child({ component: "StorageCache" });
 
-type StorageSlot = {
+export type StorageSlot = {
```
Comment (aditiharini) [inline apps/hubble/src/storage/stores/storageCache.ts:62]: Why do we want to expose a type that makes legacy_units vs units visible to the rest of the codebase? Is there a reason that other parts of the code need to know about this or can we just expose total_units via a function? I'm somewhat nervous about mishandling this legacy_units vs units distinction somewhere. I think mixing up units with legacy_units is easy to do and easy to miss in code review.

File: apps/hubble/src/storage/stores/userDataStore.ts [hunk 6]
```diff
@@ -45,7 +37,7 @@ import { messageDecode } from "../../storage/db/message.js";
  */
 class UserDataStore extends RustStoreBase<UserDataAddMessage, never> {
   constructor(db: RocksDB, eventHandler: StoreEventHandler, options: StorePruneOptions = {}) {
-    const pruneSizeLimit = options.pruneSizeLimit ?? getDefaultStoreLimit(StoreType.USER_DATA);
+    const pruneSizeLimit = options.pruneSizeLimit ?? 0;
```
Comment (aditiharini) [inline apps/hubble/src/storage/stores/userDataStore.ts:40]: I'm confused about all the places where some default value went to 0.

File: apps/hubble/src/rpc/test/server.test.ts [hunk 7]
```diff
@@ -63,7 +65,10 @@ beforeEach(async () => {
   const custodySignerKey = (await custodySigner.getSignerKey())._unsafeUnwrap();
   custodyEvent = Factories.IdRegistryOnChainEvent.build({ fid }, { transient: { to: custodySignerKey } });
   signerEvent = Factories.SignerOnChainEvent.build({ fid }, { transient: { signer: signerKey } });
-  storageEvent = Factories.StorageRentOnChainEvent.build({ fid }, { transient: { units: 1 } });
+  storageEvent = Factories.StorageRentOnChainEvent.build(
```
Comment (aditiharini) [inline apps/hubble/src/rpc/test/server.test.ts:68]: Should we add a test that has both a legacy unit and a 2024 unit?

**Comment by aditiharini** [comment]
This seems like a pretty involved PR. What's the rollout story? Are we going to rollout to all of our hubs at the same time? Are we going to roll out to one at a time and observe that they behave as expected?

File: apps/hubble/src/rpc/test/server.test.ts (hunk 7 above)
Comment (sanjayprabhu) [reply apps/hubble/src/rpc/test/server.test.ts:68]: Fixed.

**Pull request merged**